File: apps/web/src/app/monitoring/sentry.ts
import type { App } from 'vue';
import type { Router } from 'vue-router';

function scrubEventPayload(payload: unknown): unknown {
  if (Array.isArray(payload)) {
    return payload.map(scrubEventPayload);
  }
  if (payload && typeof payload === 'object') {
    return Object.fromEntries(
      Object.entries(payload).map(([key, value]) => {
        const lowered = key.toLowerCase();
        if (
          ['authorization', 'password', 'token', 'content', 'prompt', 'secret', 'jwt'].some((token) =>
            lowered.includes(token),
          )
        ) {
          return [key, '[Filtered]'];
        }
        return [key, scrubEventPayload(value)];
      }),
    );
  }
  return payload;
}

function getSampleRate(): number {
  const raw = Number(import.meta.env.VITE_SENTRY_TRACES_SAMPLE_RATE || '0');
  return Number.isFinite(raw) ? raw : 0;
}

export async function initBrowserSentry(app: App, router: Router) {
  const dsn = import.meta.env.VITE_SENTRY_DSN_WEB;
  if (!dsn) {
    return;
  }

  const Sentry = await import('@sentry/vue');
  Sentry.init({
    app,
    dsn,
    environment: import.meta.env.VITE_SENTRY_ENVIRONMENT || 'development',
    integrations: [Sentry.browserTracingIntegration({ router })],
    tracesSampleRate: getSampleRate(),
    beforeSend(event) {
      return scrubEventPayload(event) as typeof event;
    },
    sendDefaultPii: false,
  });
}

File: apps/api/app/core/sentry.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from app.core.config import get_settings

SENSITIVE_KEYS = {
    "authorization",
    "password",
    "token",
    "jwt",
    "secret",
    "content",
    "content_json",
    "prompt",
}


def _scrub(value: Any) -> Any:
    if isinstance(value, Mapping):
        cleaned: dict[str, Any] = {}
        for key, nested in value.items():
            lowered = str(key).lower()
            if any(token in lowered for token in SENSITIVE_KEYS):
                cleaned[str(key)] = "[Filtered]"
            else:
                cleaned[str(key)] = _scrub(nested)
        return cleaned
    if isinstance(value, list):
        return [_scrub(item) for item in value]
    if isinstance(value, str) and len(value) > 2000:
        return value[:2000]
    return value


def sentry_before_send(event: dict[str, Any], hint: dict[str, Any]) -> dict[str, Any]:
    _ = hint
    return _scrub(event)


def init_sentry_api() -> None:
    settings = get_settings()
    if not settings.sentry_dsn_api:
        return

    try:
        import sentry_sdk
        from sentry_sdk.integrations.fastapi import FastApiIntegration
    except ImportError:
        return

    sentry_sdk.init(
        dsn=settings.sentry_dsn_api,
        environment=settings.sentry_environment,
        traces_sample_rate=settings.sentry_traces_sample_rate_api,
        before_send=sentry_before_send,
        integrations=[FastApiIntegration()],
        send_default_pii=False,
    )
